modules/dreamview/backend/point_cloud/point_cloud_updater.cc

```diff
@@ -0,0 +1,90 @@
+ * Copyright 2018 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/dreamview/backend/point_cloud/point_cloud_updater.h"
+
+#include "modules/common/adapters/adapter_manager.h"
+#include "modules/common/log.h"
+#include "modules/common/time/time.h"
+#include "modules/dreamview/backend/common/dreamview_gflags.h"
+#include "pcl_conversions/pcl_conversions.h"
+#include "pcl/point_cloud.h"
+#include "pcl/point_types.h"
+#include "third_party/json/json.hpp"
+
+namespace apollo {
+namespace dreamview {
+
+using apollo::common::adapter::AdapterManager;
+using apollo::common::time::Clock;
+using sensor_msgs::PointCloud2;
+using Json = nlohmann::json;
+
+PointCloudUpdater::PointCloudUpdater(WebSocketHandler *websocket)
+    : websocket_(websocket) {
+  RegisterMessageHandlers();
+}
+
+void PointCloudUpdater::RegisterMessageHandlers() {
+  websocket_->RegisterMessageHandler(
+      "RequestPointCloud",
+      [this](const Json &json, WebSocketHandler::Connection *conn) {
+        std::string to_send;
+        // If there is no point_cloud data for more than 2 seconds, reset.
+        if (Clock::NowInSeconds() - last_receive_time_ > 2.0) {
+          point_cloud_.Clear();
+          boost::unique_lock<boost::shared_mutex> writer_lock(mutex_);
+          point_cloud_.SerializeToString(&point_cloud_str_);
+        }
+        {
+          boost::shared_lock<boost::shared_mutex> reader_lock(mutex_);
+          to_send = point_cloud_str_;
+        }
+        websocket_->SendBinaryData(conn, to_send, true);;
+      });
+}
+
+void PointCloudUpdater::Start() {
+  AdapterManager::AddPointCloudCallback(&PointCloudUpdater::UpdatePointCloud,
+                                        this);
+}
+
+void PointCloudUpdater::UpdatePointCloud(const PointCloud2 &point_cloud) {
+  last_receive_time_ = Clock::NowInSeconds();
+  // transform from ros to pcl
+  pcl::PointCloud<pcl::PointXYZ> pcl_data;
+  pcl::fromROSMsg(point_cloud, pcl_data);
+
+  point_cloud_.Clear();
+  int count = 0;
+  for (size_t idx = 0; idx < pcl_data.size(); idx += 4) {
+    pcl::PointXYZ& pt = pcl_data.points[idx];
+    if (!isnan(pt.x) && !isnan(pt.y) && !isnan(pt.z)) {
+      count++;
+      point_cloud_.add_num(pt.x);
+      point_cloud_.add_num(pt.y);
+      point_cloud_.add_num(pt.z + 1.91);
```

Comment:
Are we able to get the magic number from some config? Different car model would have different height.

Reply:
@unacao @siyangy Is that case put it in https://github.com/ApolloAuto/apollo/blob/master/modules/common/configs/proto/vehicle_config.proto maybe? This is where we usually load vehicle related configs.

Reply:
looks like it's from velodyne64_height.yaml, not sure we have a programatic way to read it though.